test/ruby/test_shapes.rb

```diff
@@ -226,6 +226,34 @@ def initialize
     end;
   end
 
+  def test_run_out_of_shape_rb_obj_copy_ivar
+    assert_separately([], "#{<<~"begin;"}\n#{<<~'end;'}")
+    begin;
+      class A
+        def initialize
+          init # Avoid right sizing
+        end
+
+        def init
+          @a = @b = @c = @d = @e = @f = 1
+        end
+      end
+
+      a = A.new
+
+      o = Object.new
+      i = 0
+      while RubyVM::Shape.shapes_available > 1
+        o.instance_variable_set(:"@i#{i}", 1)
+        i += 1
+      end
+
+      GC.disable # FIXME: if GC is enabled we run into another bug
+
+      a.dup
```

Comment:
So here we create a first instance of `A` that wasn't right sized and had to spill over in a malloc.

Then after running out of shapes we dup it, Ruby tries to rebuild the correct shape tree and run out of shapes in the middle of the process.